File: Python/Remove Duplicates from Sorted List II.py
# Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

# For example,
# Given 1->2->3->3->4->4->5, return 1->2->5.
# Given 1->1->1->2->3, return 2->3.


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
		if head == None:
			return None
		safeG = ListNode(-65536)
		safeG.next = head
		current = safeG
		pre = head
		while pre != None:
			isDuplicate = False
			while pre.next != None and pre.val == pre.next.val:
				isDuplicate = True
				temp = pre
				pre = pre.next
				
			if isDuplicate:
				temp = pre
				pre = pre.next
				continue
			current.next = pre
			current = current.next
			pre = pre.next
		current.next = pre
		temp = safeG.next
		return temp

File: Java/Remove Duplicates from Sorted List II.java
class Remove Duplicates from Sorted List II{
	public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode curr = head;
        while(curr != null && curr.next != null){
        	if(curr.val != curr.next.val){
        		curr=curr.next;
        		pre = pre.next;
        	}
        	else{
        		while(curr.next != null && curr.val == curr.next.val){
        			curr.next = curr.next.next;
        		}
        		pre.next = curr.next;
        		curr = pre.next;
        	}
        }
        return dummy.next;
    }
} 